corehq/apps/reports/standard/sms.py

```diff
@@ -1097,3 +1105,191 @@ def shared_pagination_GET_params(self):
+
+    def _fmt_row(self, number, owner_cache, link_user):
+        if isinstance(number, PhoneNumber):
+            return [
+                self._fmt_owner(number.owner_doc_type, number.owner_id, owner_cache, link_user),
+                number.phone_number,
+                self._fmt_status(number),
+                "Yes" if number.is_two_way else "No",
+            ]
+
+        return [
+            self._fmt_owner(number.owner_doc_type, number.owner_id, owner_cache),
+            '---',
+            '---',
+            '---',
+        ]
+
+    def _get_rows(self, paginate=True, link_user=True):
+        owner_cache = {}
+        if self._show_users_without_phone_numbers:
+            data = self._get_users_without_phone_numbers()
+        else:
+            data = self._get_queryset()
+
+        if paginate and self.pagination:
+            data = data[self.pagination.start:self.pagination.start + self.pagination.count]
+
+        for number in data:
+            yield self._fmt_row(number, owner_cache, link_user)
+
+    def _get_queryset(self):
+        query = PhoneNumber.objects.filter(domain=self.domain)
+
+        if self.filter_type == 'phone_number':
+            if self.phone_number_filter:
+                query = query.filter(phone_number=self.phone_number_filter)
+        elif self.filter_type == 'contact':
+            if self._show_cases:
+                query = query.filter(owner_doc_type='CommCareCase')
+            else:
+                query = query.filter(owner_doc_type__in=['CommCareUser', 'WebUser'])
+                if self.selected_group:
+                    query = query.filter(owner_id__in=self.user_ids_in_selected_group)
+
+            if self.verification_status == 'not_verified':
+                query = query.filter(pending_verification=False, verified=False)
+            elif self.verification_status == 'verification_pending':
+                query = query.filter(pending_verification=True)
+            elif self.verification_status == 'verified':
+                query = query.filter(verified=True)
+
+        return query.order_by('phone_number', 'couch_id')
+
+    @memoized
+    def _get_users_without_phone_numbers(self):
+        query = (
+            PhoneNumber.objects.filter(domain=self.domain).filter(owner_doc_type__in=['CommCareUser', 'WebUser'])
+        )
+
+        if self.selected_group:
+            users_by_id = {
+                id: {'_id': id, 'doc_type': 'CommCareUser'}
+                for id in self.user_ids_in_selected_group
+            }
+            query.filter(owner_id__in=users_by_id.keys())
+        else:
+            users_by_id = {u['id']: u for u in get_user_id_and_doc_type_by_domain(self.domain)}
+
+        user_ids_with_phone_numbers = {x[0] for x in query.values_list('owner_id').distinct()}
+        user_ids = set(users_by_id.keys()) - user_ids_with_phone_numbers
+        user_types_with_id = {(id, users_by_id[id]['doc_type']) for id in user_ids}
```

Comment:
i think this can just be a list, right? might perform a bit better as a list since it wouldn't have to keep track of set-related info